src/protocols/secure_channel/CASESession.cpp

```diff
@@ -339,57 +309,62 @@ CHIP_ERROR CASESession::DeriveSecureSession(SecureSession & session, SecureSessi
+#ifdef ENABLE_HSM_CASE_EPHEMERAL_KEY
+    mEphemeralKey.SetKeyId(CASE_EPHEMERAL_KEY);
+#endif
+    ReturnErrorOnFailure(mEphemeralKey.Initialize());
 
-    // Step 1
     // Fill in the random value
     ReturnErrorOnFailure(DRBG_get_bytes(initiatorRandom, kSigmaParamRandomNumberSize));
 
-// Step 4
-#ifdef ENABLE_HSM_CASE_EPHERMAL_KEY
-    mEphemeralKey.SetKeyId(CASE_EPHEMERAL_KEY);
-#endif
-    ReturnErrorOnFailure(mEphemeralKey.Initialize());
+    // Construct Sigma1 Msg
+    msg_R1 = System::PacketBufferHandle::New(data_len);
+    VerifyOrReturnError(!msg_R1.IsNull(), CHIP_ERROR_NO_MEMORY);
 
-    // Start writing TLV
     tlvWriter.Init(std::move(msg_R1));
     ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
-    ReturnErrorOnFailure(tlvWriter.PutBytes(CASETLVTag::kRandom, initiatorRandom, sizeof(initiatorRandom)));
-
-    // Step 5
-    uint16_t n_trusted_roots = mOpCredSet->GetCertCount();
-    // Initiator's session ID
-    ReturnErrorOnFailure(tlvWriter.Put(CASETLVTag::kSessionID, mConnectionState.GetLocalKeyID(), true));
-    // Step 2/3
-    ReturnErrorOnFailure(tlvWriter.Put(CASETLVTag::kNumberofTrustedRootIDs, n_trusted_roots, true));
-    for (uint16_t i = 0; i < n_trusted_roots; ++i)
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(1), initiatorRandom, sizeof(initiatorRandom)));
+    // Retrieve Session Identifier
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(2), mConnectionState.GetLocalKeyID(), true));
+    // Generate a Destination Identifier
     {
-        CertificateKeyId trustedRootId = mOpCredSet->GetTrustedRootId(i);
-        if (!trustedRootId.empty())
-        {
-            ReturnErrorOnFailure(tlvWriter.PutBytes(CASETLVTag::kTrustedRootID, trustedRootId.data(), trustedRootId.size()));
-        }
+        const ChipCertificateData * rootCertificate = mOpCredSet->GetRootCertificate(mTrustedRootId);
+        VerifyOrReturnError(rootCertificate != nullptr, CHIP_ERROR_CERT_NOT_FOUND);
+        VerifyOrReturnError(!rootCertificate->mPublicKey.empty(), CHIP_ERROR_INTERNAL);
+        VerifyOrReturnError(rootCertificate->mPublicKey.size() == kP256_PublicKey_Length, CHIP_ERROR_INTERNAL);
+        ChipCertificateData nodeOperationalCertificate;
+        FabricId fabricId;
+        MutableByteSpan destinationIdSpan(destinationIdentifier, sizeof(destinationIdentifier));
+
+        ReturnErrorOnFailure(DecodeChipCert(mOpCredSet->GetDevOpCred(mTrustedRootId), mOpCredSet->GetDevOpCredLen(mTrustedRootId),
+                                            nodeOperationalCertificate));
+        ReturnErrorOnFailure(nodeOperationalCertificate.mSubjectDN.GetCertFabricId(fabricId));
+        // retrieve Fabric IPK
+        MutableByteSpan ipkSpan(mIPK, sizeof(mIPK));
+        ReturnErrorOnFailure(RetrieveIPK(fabricId, ipkSpan));
+        ReturnErrorOnFailure(GenerateDestinationID(ByteSpan(initiatorRandom, sizeof(initiatorRandom)), rootCertificate->mPublicKey,
```

Comment:
There exists a span constructor over static arrays, such that:

```
uint8_t initiatorRandom[kSigmaParamRandomNumberSize] = { 0 };

// The following two are equivalent:
ByteSpan span1(initiatorRandom);
ByteSpan span2(&initiatorRandom[0], sizeof(initiatorRandom));
```

Suggest using that version for arguments that are backed by arrays

Reply:
Updated